### PR TITLE
Feature/drmm tutorial

### PR DESCRIPTION
**Pull request by faneshion**

### CODE REVIEW

File: matchzoo/data_generator/__init__.py [hunk 1]
```diff
@@ -3,3 +3,5 @@
 from .dynamic_data_generator import DynamicDataGenerator
 from .dpool_data_generator import DPoolDataGenerator
 from .dpool_data_generator import DPoolPairDataGenerator
+from .histogram_data_generator import HistogramDataGenerator
+from .histogram_data_generator import HistogramPairDataGenerator
```
Comment (uduse) [inline matchzoo/data_generator/__init__.py]: Need an empty line at the end of the file.

File: matchzoo/data_generator/histogram_data_generator.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                 hist_mode: str = 'CH',
+                 batch_size: int = 32,
+                 shuffle: bool = True):
+        """:class:`HistogramDataGenerator` constructor."""
+        self._match_hist_unit = MatchingHistogramUnit(
+            hist_bin_size=hist_bin_size,
+            embedding_matrix=embedding_matrix,
+            normalize=True,
+            mode=hist_mode)
+        # Here the super().__init_ must be after the self._data_pack
```
Comment (uduse) [inline matchzoo/data_generator/histogram_data_generator.py]: Why?

File: matchzoo/embedding.py [hunk 3]
```diff
@@ -59,22 +59,20 @@ def build_matrix(
         self,
         term_index: typing.Union[
             dict, processor_units.VocabularyUnit.TermIndex],
-        initializer=lambda: np.random.uniform(-0.2, 0.2)
+        initialize_range: tuple = (-0.2, 0.2)
```
Comment (uduse) [inline matchzoo/embedding.py]: It's not always a uniform distribution. I usually use trunc normal, and I wanted to keep that flexibility.

File: matchzoo/data_generator/histogram_data_generator.py (hunk 2 above)
Comment (faneshion) [reply matchzoo/data_generator/histogram_data_generator.py]: emmm... This line is a copy from the `pair_data_generator`. I removed it.

File: matchzoo/embedding.py (hunk 3 above)
Comment (faneshion) [reply matchzoo/embedding.py]: To pass an initializer, we would assign each element of the matrix, as is shown in lines 75-77 in original `embedding.py`. This is too much time-consuming.

Comment (faneshion) [reply matchzoo/embedding.py]: To pass an initializer, we would assign each element of the matrix, as is shown in lines 75-77 in original embedding.py. This is too much time-consuming.

File: matchzoo/processor_units/processor_units.py [hunk 4]
```diff
@@ -531,3 +531,64 @@ def transform(self, tokens: list) -> list:
...
+
+    """
+
+    def __init__(self, hist_bin_size: int = 30, embedding_matrix=None,
+                 normalize=True, mode: str = 'LCH'):
+        """The constructor."""
+        self._hist_bin_size = hist_bin_size
+        self._embedding_matrix = embedding_matrix
+        if normalize:
+            self._embedding_matrix = self.normalize_embedding(embedding_matrix)
```
Comment (bwanglzu) [inline matchzoo/processor_units/processor_units.py]: `normalize_mebedding` is a classmethod, can you use `self`? not sure why use classmethod.

File: matchzoo/processor_units/processor_units.py [hunk 5]
```diff
@@ -531,3 +531,64 @@ def transform(self, tokens: list) -> list:
...
+        self._mode = mode
+
+    @classmethod
+    def normalize_embedding(cls,
+                            embedding_matrix: np.ndarray = None) -> np.ndarray:
+        """
+        Normalize the embedding matrix.
+
+        :param embedding: The input embedding matrix.
```
Comment (bwanglzu) [inline matchzoo/processor_units/processor_units.py]: :param embedding_matrix:

File: matchzoo/embedding.py (hunk 3 above)
Comment (faneshion) [reply matchzoo/embedding.py]: I have recovered the embedding initialization method to keep the flexibility.

File: matchzoo/models/drmm.py [hunk 6]
```diff
@@ -82,19 +96,21 @@ def attention_layer(cls, attention_input: typing.Any,
         """
         # shape = [B, L, 1]
         dense_input = keras.layers.Dense(1, use_bias=False)(attention_input)
-        if attention_mask:
+        if attention_mask is not None:
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/models/drmm.py]: It is not allowed by keras since the `attention_mask` is a keras `Tensor`.
